### DISCUSSION TITLE
RFC: Concept of marking a workspace as dirty

### DISCUSSION THREAD

**Discussion by MrPeacockNLB** [coder/coder#9477] [RFCs]
# Description

We are using ephemeral workspaces and open for each and every task a new workspace. If one is working on multiple projects simultaneously the list can get easely longer. Each workspace is derived from a repository as everybody nowadays works with a **CVS** like git. I often open a repository for read only in coder. So there were workspaces with no changes in it. So far so good. But it would be nice if one can see if there is unsaved work before deleting a workspace.

Currently one has to open the workspace and do some manual checking if there were some unsaved change in it. This is a time wasting task. Deleting a workspace means always the anwsering the security dialog or closing the workspace with the coder binary.

# Solution

If there is some unsaved work in the workspace then this workspace is marked as **dirty** in the UI. This should be periodly requested from the coder daemon like metadata values. If the executed code return a code 0 then the workspace is clean or something like that. The executed code can check if there is some unsaved work in the filesystem in the workspace.

If a workspace is marked as clean a simple delete should remove a workspace without any dialog. There should also be the ability to cleanup all clean workspaces at once.

Deleting a dirty workspace should raise the known security dialog to prevent accidentally deleting unsaved work.

**Comment by MrPeacockNLB** [comment]
I've been playing around a bit and implemented a very simple POC for this. I put some Bash scripts into a metadata resource to get current status of opened workspace.

![image](https://github.com/coder/coder/assets/103434584/8633c297-4cc2-4be5-a668-90466e51b7c5)

![image](https://github.com/coder/coder/assets/103434584/20973b7f-019f-495e-968e-20bf03b3f037)

![image](https://github.com/coder/coder/assets/103434584/24842fce-130f-4a86-8c7c-ca6e2db25c4c)

```
metadata {
    display_name = "Workspace status"
    key          = "dirty"
    script = <<EOT
      # jump into standard dir
      cd /workspace/workdir

      # check uncommitted changes
      DIFF=$(git diff --exit-code)
      if [[ $? -gt 0 ]]; then
          echo "uncommitted changes"
          exit 1
      fi

      # check for untracked files
      if output=$(git status --porcelain) && [ -z "$output" ]; then
        echo "workspace is clean"
      else
        echo "untracked files"
        exit 1
      fi
    EOT
    interval = 10
    timeout = 5
  }
```

PS: Code is far from perfect!

**Reply by matifali** [reply]
This is a good POC. Something to add to my personal templates. Thanks.

**Reply by michaelbrewer** [reply]
Awesome idea, i think having a script that can take the WORKSPACE_DIR (which might not be the same as the HOME dir), and then check for any folder that might be repos. Before doing the checks ie:

WORKSPACE_DIR=/projects has proj1, proj2, and proj3 checked out.

**Reply by michaelbrewer** [reply]
The next part which would be nice is to allow users to "Delete" workspaces which are clean via a filter on the "metadata"

**Reply by michaelbrewer** [reply]
You can also detect if changes have not been pushed yet.

```bash
      if output=$(git status --porcelain) && [ -z "$output" ]; then
        if git log --branches --not --remotes | grep . > /dev/null; then
          echo "changes not pushed"
          exit 1
        else
          echo "workspace is clean"
        fi
      else
        echo "untracked files"
        exit 1
      fi
```

**Reply by michaelbrewer** [reply]
A more complete snippet. This also handles remote changes.

```tf
metadata {
    display_name = "Workspace status"
    key          = "7_dirty"
    script       = <<EOT
      if [ -z "${local.git_url}" ]; then
        echo "not a git repo"
        exit 0
      fi

      cd ${local.workspace_home}/${local.folder_name}

      DIFF=$(git diff --exit-code)
      if [[ $? -gt 0 ]]; then
        echo "uncommitted changes"
        exit 1
      fi
      if output=$(git status --porcelain) && [ -n "$output" ]; then
        echo "untracked files"
        exit 1
      fi
      if git log --branches --not --remotes | grep . > /dev/null; then
        echo "changes not pushed"
        exit 1
      fi

      BNAME=$(git rev-parse --abbrev-ref HEAD)
      git fetch -q origin $BNAME || {
        echo "workspace is clean"
        exit 0
      }

      LOCAL=$(git rev-parse @)
      REMOTE=$(git rev-parse @{u})
      BASE=$(git merge-base @ @{u})
      if [ $LOCAL = $REMOTE ]; then
        echo "workspace is clean"
      elif [ $LOCAL = $BASE ]; then
        echo "needs to pull changes"
        exit 1
      else
        echo "workspace has diverged"
        exit 1
      fi
    EOT
    interval     = 10
    timeout      = 5
  }
```
